### ISSUE TITLE
Add CI testing for build shell scripts

### ISSUE THREAD

**Issue by chrissimpkins** [source-foundry/Hack#269]
Add Travis CI testing for the new build.sh shell script that builds Hack fonts from UFO source.

Recommended by @mavit in https://github.com/source-foundry/Hack/issues/227#issuecomment-322526873

Testing tool suggested: https://www.shellcheck.net/

Travis CI implementation: https://github.com/koalaman/shellcheck/wiki/TravisCI

TODO:

- [x] implement shellcheck testing on `build-web.sh` script
- [x] implement shellcheck testing on Werner's ttfautohint install script
- [x] implement shellcheck testing on web font build scripts

**Comment by mavit** [comment]
Looks like this is now implemented on branch `dev-build-scripts-shellCItest-patch`?

**Comment by chrissimpkins** [comment]
Yes and seems to be working well.  I really appreciate the suggestion.  Sorry haven't had a chance to update the issue report yet.  Was on the agenda for this weekend.  

I do have a question for you if you have experience with Travis.  I would prefer to implement this concurrent to the ufolint tests using their matrix build process.  In the branch you mentioned, I have the test in line (i.e. sequential) with our ufo source tests.  Do you know of a way to split this out to a separate matrix 'build' using Travis terminology so that it runs as an isolated test independent from the ufo source tests?  I am aiming to do this with the build.sh test that is a simple execution of the build.sh script with the dependency install flag.  

Note currently failing in that branch because of my failed attempts to install the harfbuzz dependency.  Werner Lemburg contributed a script to take care of these elements and once I have this in place over the weekend we should be passing tests again.

**Comment by chrissimpkins** [comment]
merged into `dev-build-scripts` branch while will remain remote working branch for the build script work

https://github.com/source-foundry/Hack/commit/19a2b9390e49d7db93a90e04e05fe51b3e79b187

**Comment by chrissimpkins** [comment]
In addition to shellcheck tests of build.sh script, the test now includes complete script execution with dependency installs for fontmake, fontTools, and ttfautohint.  Passes tests on Travis as of https://github.com/source-foundry/Hack/commit/19a2b9390e49d7db93a90e04e05fe51b3e79b187

**Comment by mavit** [comment]
> In the branch you mentioned, I have the test in line (i.e. sequential) with our ufo source tests. Do you know of a way to split this out to a separate matrix 'build' using Travis terminology so that it runs as an isolated test independent from the ufo source tests?

It's perhaps not especially elegant, but https://github.com/mavit/Hack/commit/8b28b0da7c2034ddebaa6a9a3b459c2c83fda88b does the trick.

An alternative approach would be to wrap Make around the build process, as in the example at https://github.com/mavit/Hack/commit/5e0b8f319bc76f1d3f6313198c3113fe52286dd1.  If you're not familiar with Make, it looks old fashioned but it's actually surprisingly powerful.  It understands the build process as a collection of rules, "to build this file, I must first build these files, then I must run these commands".  Consequently, it's able to run things in parallel where possible, and skip files that haven't changed since the last build.

**Comment by chrissimpkins** [comment]
:+1:  fantastic!  I love both of these approaches.  For the make approach in this portion of the makefile: 

```
build: source/*.ufo
	./build.sh

build-with-dependencies: source/*.ufo
	./build.sh --install-dependencies

lint: shellcheck ufolint

shellcheck: build.sh
	$@ $^

ufolint: source/*.ufo
	$@ $^
```

can you explain what the combined idioms: `$@ $^` and do?  `$@` is the argument list to the script but I am not familiar with `$^`.   It appears that this is a way to execute the commands `shellcheck build.sh` and `ufolint source/*.ufo` from those build dependencies defined in the `lint` line?  Is it possible to extend the idiom you are using here with wildcards?  For instance, could we convert the `build.sh` file in the `shellcheck` definition to include all shell scripts in the repository (which we will define with `*.sh` extensions?

**Comment by chrissimpkins** [comment]
I created PR https://github.com/source-foundry/Hack/pull/274 with the make file testing changes from @mavit.  Peter's commit currently in the `mavit-make` branch (working branch off of https://github.com/source-foundry/Hack/tree/dev-build-scripts).

Peter let's discuss the issues in https://github.com/source-foundry/Hack/issues/269#issuecomment-323590660 before merge into the `dev-build-scripts` branch.

**Comment by chrissimpkins** [comment]
Docs ftw!

`$@` = rule target
`$^` = space delimited list of names of all pre-requisites for target

To add additional files to shellcheck, we will modify the pre-requisites for the `shellcheck:` target (can use wildcards to capture all shell scripts when the old ones are cleared out of repo, pre-req wildcards are expanded with this approach as used for the `source/*.ufo` wildcard)

**Comment by chrissimpkins** [comment]
makefile testing approach is now merged into `dev-build-scripts` branch
